### PR TITLE
New feature and improvements

### PR DESCRIPTION
**Pull request by Apceniy**
Added a new feature to the remind command when you can see active reminders

### CODE REVIEW

File: src/main/java/ru/l1ttleO/boyfriend/Actions.java [hunk 1]
```diff
@@ -106,16 +101,11 @@ public static void muteMember(final @NotNull MessageChannel channel, final @NotN
...
-                try {
-                    Thread.sleep(duration * 1000L);
-                    unmuteMember(null, role, guild.getSelfMember(), muted, "Время наказания истекло", silent);
-                } catch (final InterruptedException ignored) {
-                }
-            }, "Mute timer " + muted.getId());
-            guildMutes.put(muted.getIdLong(), thread);
+            final DelayedRunnable runnable = new DelayedRunnable(MUTES_THREAD_GROUP, (DelayedRunnable thisDR) -> {
+                unmuteMember(null, role, guild.getSelfMember(), muted, "Время наказания истекло", silent);
+            }, "Mute timer " + muted.getId(), duration * 1000L, null);
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Actions.java:106]:
```suggestion
            final DelayedRunnable runnable = new DelayedRunnable(MUTES_THREAD_GROUP, (DelayedRunnable thisDR) -> unmuteMember(null, role, guild.getSelfMember(), muted, "Время наказания истекло", silent),
                                                                 "Mute timer " + muted.getId(), duration * 1000L, null);
```
Reason: `Statement lambda can be replaced with expression lambda`

File: src/main/java/ru/l1ttleO/boyfriend/DelayedRunnable.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+import java.util.function.Consumer;
+import javax.annotation.Nullable;
+import org.jetbrains.annotations.NotNull;
+
+public class DelayedRunnable {
+
+    final public Thread thread;
+    final public long duration;
+    final public long startedAt;
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/DelayedRunnable.java:11]:
```suggestion
    public final Thread thread;
    public final long duration;
    public final long startedAt;
```
Reason: `Missorted modifiers`

File: src/main/java/ru/l1ttleO/boyfriend/DelayedRunnable.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    public DelayedRunnable(final @Nullable ThreadGroup threadGroup, final @NotNull Consumer<DelayedRunnable> runnable, final @Nullable String name, final @NotNull long millis, final @Nullable Consumer<DelayedRunnable> ifInterrupted) {
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/DelayedRunnable.java:13]:
```suggestion
    public DelayedRunnable(@Nullable final ThreadGroup threadGroup, final @NotNull Consumer<DelayedRunnable> runnable, @Nullable final String name, final long millis, @Nullable final Consumer<DelayedRunnable> ifInterrupted) {
```
Reason: `Missorted modifiers`, `Primitive type members cannot be annotated`

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 4]
```diff
@@ -55,16 +57,41 @@ public static String getBeep() {
...
+        }
+
+        int result = 0;
+        final Set<String> used = Collections.emptySet();
+        String[] buffer;
+        final StringBuilder input = new StringBuilder(toParse);
+        while (!input.isEmpty()) {
+            buffer = input.toString().split("\\d+$");
+            if (buffer.length < 2 || used.contains(buffer[1]))
+                throw new NumberFormatException();
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:73]: `'new NumberFormatException()' without arguments`

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 5]
```diff
@@ -55,16 +57,41 @@ public static String getBeep() {
...
 
-    public static int getDurationMultiplied(final @NotNull String toParse) {
+    public static int getDurationMultiplied(final @NotNull String toParse) throws NumberFormatException {
         try {
+            return Integer.parseInt(toParse);
+        } catch (final @NotNull NumberFormatException ignored) {
+        }
+
+        int result = 0;
+        final Set<String> used = Collections.emptySet();
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:67]:
```suggestion
        final Set<String> used = new java.util.HashSet<>(Collections.emptySet());
```
Reason: Fixes `Immutable object is modified: line 74` and `Condition 'used.contains(buffer[1])' is always 'false' when reached: line 72`

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 6]
```diff
@@ -55,16 +57,41 @@ public static String getBeep() {
...
+            if (buffer.length < 2 || used.contains(buffer[1]))
+                throw new NumberFormatException();
+            used.add(buffer[1]);
+            input.setLength(input.length() - buffer[1].length());
+
             int multiplier = 1;
-            for (int i = 0; i < DURATION_TEXTS.length; i++) {
-                if (toParse.endsWith(DURATION_TEXTS[i][0])) return multiplier * Integer.parseInt(toParse.substring(0, toParse.length() - 1));
+            int i = 0;
+            for (; i < DURATION_TEXTS.length; i++) {
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:79]:
```suggestion
            int i;
            for (i = 0; i < DURATION_TEXTS.length; i++) {
```
Reason: `'for' statement lacks initializer`

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 7]
```diff
@@ -55,16 +57,41 @@ public static String getBeep() {
...
+                if (buffer[0].equals(DURATION_TEXTS[i][0]))
+                    break;
                 multiplier *= DURATION_MULTIPLIERS[i];
             }
-            return Integer.parseInt(toParse);
-        } catch (final @NotNull NumberFormatException ignored) {
-            return 0;
+            if (i == DURATION_TEXTS.length) // modifier not found
+                throw new NumberFormatException();
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:85]: `'new NumberFormatException()' without arguments`

File: src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java [hunk 8]
```diff
@@ -1,43 +1,79 @@
...
+    public static final @NotNull HashMap<Long, HashMap<DelayedRunnable, String>> REMINDERS = new HashMap<>();
+
     public void run(final @NotNull MessageReceivedEvent event, final @NotNull String @NotNull [] args) throws InvalidAuthorException, WrongUsageException {
         final int duration;
         final Member author = event.getMember();
         final MessageChannel channel = event.getChannel();
         final String text;
-        if (args.length < 3)
-            throw new WrongUsageException("Требуется указать текст напоминания!", channel, this.getUsages());
+        if (args.length < 2) {
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java:30]:
```suggestion
        if (author == null)
            throw new InvalidAuthorException();
        if (args.length < 2) {
```
Reason: `Method invocation 'getIdLong' may produce 'NullPointerException': line 33`

File: src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java [hunk 9]
```diff
@@ -1,43 +1,79 @@
...
+            else {
+                final StringBuilder listText = new StringBuilder("Активные напоминания:");
+                String append;
+                for (final Entry<DelayedRunnable, String> reminder : userReminders.entrySet()) {
+                    append = """
+                        \n
+                        <t:%s:R>
+                        %s""".formatted((reminder.getKey().startedAt + reminder.getKey().duration) / 1000, reminder.getValue());
+                    if (listText.isEmpty())
+                        append.strip();
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java:43]: `Result of 'String.strip()' is ignored`

File: src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java [hunk 10]
```diff
@@ -1,43 +1,79 @@
...
+                    if (listText.length() + append.length() > 2000) {
+                        channel.sendMessage(listText).queue();
+                        listText.setLength(0); // clear
+                        if (append.length() > 2000)
+                            append = append.substring(0, 1994) + "..." + append.substring(append.length() - 3, append.length());
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java:48]:
```suggestion
                            append = append.substring(0, 1994) + "..." + append.substring(append.length() - 3);
```
Reason: `Unnecessary string length argument`

File: src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java [hunk 11]
```diff
@@ -1,43 +1,79 @@
...
+                    }
+                    listText.append(append);
+                }
+                channel.sendMessage(listText).queue();
+            }
+            return;
+        } else if (args.length < 3)
+            throw new WrongUsageException("Требуется указать текст напоминания!");
         if (author == null)
             throw new InvalidAuthorException();
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java:58]:
```suggestion
```
Reason: Already checked above

File: src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java [hunk 12]
```diff
@@ -1,43 +1,79 @@
...
+            throw new WrongUsageException("Длительность должна быть больше нуля!");
+        text = Utils.wrap(StringUtils.join(args, ' ', 2, args.length));
+        channel.sendMessage("Напоминание успешно установлено. Через %s будет отправлено данное сообщение: %s".formatted(Utils.getDurationText(duration, true), text)).queue();
+
+        final DelayedRunnable runnable = new DelayedRunnable(REMINDERS_THREAD_GROUP, (final DelayedRunnable dr) -> {
+            channel.sendMessage(author.getAsMention() + " " + text).queue();
+            REMINDERS.getOrDefault(author.getIdLong(), new HashMap<>()).remove(dr);
+        }, "Remind timer " + author.getId(), duration * 1000L, (final DelayedRunnable dr) -> {
+            Actions.sendNotification(event.getGuild(), "Прерван таймер напоминания для %s: %s".formatted(author.getAsMention(), text), false);
+        });
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java:74]:
```suggestion
        }, "Remind timer " + author.getId(), duration * 1000L, (final DelayedRunnable dr) -> Actions.sendNotification(event.getGuild(), "Прерван таймер напоминания для %s: %s".formatted(author.getAsMention(), text), false));
```
Reason: `Statement lambda can be replaced with expression lambda`

**Review by l1ttleO — CHANGES_REQUESTED** [review]

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java (hunk 6 above)
Comment (Apceniy) [reply src/main/java/ru/l1ttleO/boyfriend/Utils.java:79]: what's the difference?

Comment (l1ttleO) [reply src/main/java/ru/l1ttleO/boyfriend/Utils.java:79]: looks less confusing i guess

File: src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java (hunk 12 above)
Comment (Apceniy) [reply src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java:74]: Maybe keep a first line break for readability?

Comment (l1ttleO) [reply src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java:74]: sure

Comment (Apceniy) [reply src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java:74]: I think you misplaced a whitespace and broke indentation

Comment (l1ttleO) [reply src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java:74]: IntelliJ thinks it's indented correctly

Comment (Apceniy) [reply src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java:74]: ok, whatever
maybe GitHub plays with me, idk

Comment (Apceniy) [reply src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java]: see what I said?

Comment (l1ttleO) [reply src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java]: image
did it not copy-paste properly?

Comment (Apceniy) [reply src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java]: > edit: fixed, you can commit this suggestion now

no, I can't
image

File: src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java [hunk 13]
```diff
@@ -1,43 +1,78 @@
...
+        if (duration < 1)
+            throw new WrongUsageException("Длительность должна быть больше нуля!");
+        text = Utils.wrap(StringUtils.join(args, ' ', 2, args.length));
+        channel.sendMessage("Напоминание успешно установлено. Через %s будет отправлено данное сообщение: %s".formatted(Utils.getDurationText(duration, true), text)).queue();
+
+        final DelayedRunnable runnable = new DelayedRunnable(REMINDERS_THREAD_GROUP, (final DelayedRunnable dr) -> {
+            channel.sendMessage(author.getAsMention() + " " + text).queue();
+            REMINDERS.getOrDefault(author.getIdLong(), new HashMap<>()).remove(dr);
+        }, "Remind timer " + author.getId(), duration * 1000L, (final DelayedRunnable dr) ->
+         Actions.sendNotification(event.getGuild(), "Прерван таймер напоминания для %s: %s".formatted(author.getAsMention(), text), false));
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java:73]:
```suggestion
                Actions.sendNotification(event.getGuild(), "Прерван таймер напоминания для %s: %s".formatted(author.getAsMention(), text), false));
```
ffs github

File: src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java (hunk 12 above)
Comment (l1ttleO) [reply src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java]: fixed

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 14]
```diff
@@ -55,16 +56,41 @@ public static String getBeep() {
...
+                    break;
                 multiplier *= DURATION_MULTIPLIERS[i];
             }
-            return Integer.parseInt(toParse);
-        } catch (final @NotNull NumberFormatException ignored) {
-            return 0;
+            if (i == DURATION_TEXTS.length)
+                throw new NumberFormatException("Time multiplier \"%s not found\"".formatted(buffer[1]));
+
+            buffer = (" " + input.toString()).split("\\D+(?=\\d+$)");
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:86]:
```suggestion
            buffer = (" " + input).split("\\D+(?=\\d+$)");
```
Reason: `Unnecessary 'toString()' call`

**Review by l1ttleO — APPROVED** [review]
LGTM, thank you for your contributions

**Pull request merged**